src/main/java/com/cloudogu/smeagol/wiki/infrastructure/PageController.java

```diff
@@ -0,0 +1,114 @@
+
+    private final PageResourceAssembler assembler = new PageResourceAssembler();
+
+    private final PageRepository repository;
+    private final WildcardPathExtractor pathExtractor;
+    private final CommandBus commandBus;
+
+    @Autowired
+    public PageController(PageRepository repository, WildcardPathExtractor pathExtractor, CommandBus commandBus) {
+        this.repository = repository;
+        this.pathExtractor = pathExtractor;
+        this.commandBus = commandBus;
+    }
+
+    @RequestMapping("**")
+    public ResponseEntity<PageResource> findByWikiIdAndPath(
+            HttpServletRequest request,
+            @PathVariable("repositoryId") String repositoryId,
+            @PathVariable("branch") String branch
+    ) {
+        WikiId id = new WikiId(repositoryId, branch);
+        Path path = pathExtractor.extractPathFromRequest(request, MAPPING, id);
+
+        Optional<Page> byWikiIdAndPath = repository.findByWikiIdAndPath(id, path);
+        if (byWikiIdAndPath.isPresent()) {
+            return ResponseEntity.ok(assembler.toResource(byWikiIdAndPath.get()));
+        }
+        return ResponseEntity.notFound().build();
+    }
+
+    @RequestMapping(method = RequestMethod.POST, value = "**")
+    public ResponseEntity<Void> createOrEdit(
+            HttpServletRequest request,
+            @PathVariable("repositoryId") String repositoryId,
+            @PathVariable("branch") String branch,
+            @RequestBody CreateOrEditRequestPayload payload
+    ) throws URISyntaxException {
+        WikiId id = new WikiId(repositoryId, branch);
+        Path path = pathExtractor.extractPathFromRequest(request, MAPPING, id);
+
+        // TODO return new page? this would safe us one request from the frontend. Is this resty?
+
+        if ( repository.exists(id, path) ) {
+            return edit(id, path, payload);
+        }
+        return create(request, id, path, payload);
+    }
+
+    private ResponseEntity<Void> create(HttpServletRequest request, WikiId id, Path path, CreateOrEditRequestPayload payload) throws URISyntaxException {
+        CreatePageCommand command = new CreatePageCommand(id, path, payload.getMessage(), payload.getContent());
+        commandBus.execute(command);
+        return ResponseEntity.created(new URI(request.getRequestURI())).build();
+    }
+
+    private ResponseEntity<Void> edit(WikiId id, Path path, CreateOrEditRequestPayload payload) {
+        EditPageCommand command = new EditPageCommand(id, path, payload.getMessage(), payload.getContent());
+        commandBus.execute(command);
+        return ResponseEntity.noContent().build();
+    }
+
+    @RequestMapping(method = RequestMethod.DELETE, value = "**")
+    public ResponseEntity<Void> delete(
+            HttpServletRequest request,
+            @PathVariable("repositoryId") String repositoryId,
+            @PathVariable("branch") String branch,
+            @RequestBody DeleteRequestPayload payload
+    ) throws URISyntaxException {
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove the declaration of thrown exception 'java.net.URISyntaxException', as it cannot be thrown from method's body. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://ecosystem.cloudogu.com/sonar/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)